lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -381,88 +412,283 @@ void GrandCentralSignalMappingsPass::runOnOperation() {
     }
   }
 
-  auto reduce = [](const Result &acc, Result result) -> Result {
-    DenseMap<FModuleOp, DenseSet<unsigned>> foo = acc.forcedInputPorts;
-    foo.insert(result.forcedInputPorts.begin(), result.forcedInputPorts.end());
-    return {acc.allAnalysesPreserved && result.allAnalysesPreserved, foo};
+  struct Result {
+    bool allAnalysesPreserved;
+    FModuleOp module;
+    SmallVector<SignalMapping> remoteMappings;
   };
-  // Note: this uses (unsigned)true instead of (bool)true for the reduction
-  // because llvm::parallelTransformReduce uses the "data" method of std::vector
-  // which is NOT provided for bool for optimization reasons.
-  Result result = llvm::parallelTransformReduce(
-      modules.begin(), modules.end(), Result(), reduce, processModule);
-
-  auto *instanceGraph = &getAnalysis<InstanceGraph>();
-  DenseMap<FModuleOp, ModuleNamespace> moduleNamespaces;
-  for (auto fixup : result.forcedInputPorts) {
-    for (auto portIdx : fixup.second) {
-      for (auto *use : instanceGraph->lookup(fixup.first)->uses()) {
+
+  // Pre-allocate result vector
+  SmallVector<Result> results;
+  results.resize(modules.size());
+
+  mlir::parallelForEachN(
+      circuit.getContext(), 0, modules.size(),
+      [this, &modules, &results](size_t index) {
+        ModuleSignalMappings mapper(modules[index], markDut, prefix);
+        mapper.run();
+        results[index] = {
+            mapper.allAnalysesPreserved, modules[index],
+            SmallVector<SignalMapping>{llvm::make_filter_range(
+                mapper.mappings, [](auto &m) { return !m.isLocal; })}};
+      });
+  bool allAnalysesPreserved =
+      llvm::all_of(results, [](auto &r) { return r.allAnalysesPreserved; });
+
+  // If this is a subcircuit, then emit JSON information necessary to drive
+  // SiFive tools.
+  // Otherwise handle any forced input ports and emit updated mappings.
+  if (isSubCircuit) {
```

Comment:
The bodies of this `if`/`else` are pretty chunky. Would this be better presented by factoring the bodies into static functions?

Reply:
They might become member functions of the pass instead, since the else branch accesses bits from the pass like `emitError`, `getAnalysis`, etc.  But yes this should be refactored, thanks!